assets/js/components/ProjectEditor.js

```diff
@@ -0,0 +1,202 @@
+import $ from 'jquery'
+import { MDCSelect } from '@material/select'
+import { ProgramEditorDialog } from '../custom/ProgramEditorDialog'
+import { showDefaultTopBarTitle, showCustomTopBarTitle } from '../layout/top_bar'
+
+export function ProjectEditor (projectDescriptionCredits, programId, textFields, showLanguageSelect, defaultText) {
+  const self = this
+  const editProgramButton = $('#edit-program-button')
+
+  this.defaultText = defaultText
+  this.programId = programId
+  this.showLanguageSelect = showLanguageSelect
+  this.textFields = textFields
+
+  this.body = $('body')
+  this.editTextUI = $('#edit-text-ui')
+  this.languageSelectorList = $('#edit-language-selector-list')
+  this.selectedLanguage = $('#edit-selected-language')
+  this.textLoadingSpinner = $('#edit-loading-spinner')
+
+  this.languageSelect = new MDCSelect(document.querySelector('#edit-language-selector'))
+  this.languages = {}
+  this.previousIndex = 0
+
+  this.closeEditorDialog = new ProgramEditorDialog(
+    projectDescriptionCredits.data('trans-close-editor-prompt'),
+    projectDescriptionCredits.data('trans-save'),
+    projectDescriptionCredits.data('trans-discard')
+  )
+
+  this.keepOrDiscardDialog = new ProgramEditorDialog(
+    projectDescriptionCredits.data('trans-save-on-language-change'),
+    projectDescriptionCredits.data('trans-keep'),
+    projectDescriptionCredits.data('trans-discard')
+  )
+
+  $('#edit-submit-button').on('click', () => { this.save() })
+
+  this.languageSelect.listen('MDCSelect:change', () => {
+    if (!this.editTextUI.is(':visible') || this.languageSelect.selectedIndex === this.previousIndex) {
+      return
+    }
+
+    if (this.areChangesSaved()) {
+      this.previousIndex = this.languageSelect.selectedIndex
+      this.getNewText()
+    } else {
+      this.keepOrDiscardDialog.show(keepOrDiscardChangesResult)
+    }
+  })
+
+  this.popStateHandler = function () {
+    if (self.areChangesSaved()) {
+      self.close()
+    } else {
+      self.closeEditorDialog.show(closeEditorDialogResult)
+    }
+  }
+
+  $(document).ready(getLanguages)
+
+  editProgramButton.on('click', () => {
+    this.show()
+  })
+
+  this.show = () => {
+    for (const textField of this.textFields) {
+      textField.show()
+    }
+
+    const langaugeSelectElement = $('#edit-language-selector')
+    if (this.showLanguageSelect) {
+      langaugeSelectElement.removeClass('d-none')
+    } else {
+      langaugeSelectElement.addClass('d-none')
+    }
+
+    window.history.pushState(
+      { type: 'ProjectEditor', id: programId, full: true },
+      $(this).text(),
+      '#' + programId
+    )
+
+    $(window).on('popstate', this.popStateHandler)
+    showCustomTopBarTitle('Program Editor', function () {
```

Comment:
Translatable text. Also is "Program Editor" a good name? what if we just leave the title empty?